src/diffusers/pipelines/pipeline_utils.py

```diff
@@ -1653,6 +1666,139 @@ def set_attention_slice(self, slice_size: Optional[int]):
+                A new pipeline with the same weights and configurations as `pipeline`.
+
+        Examples:
+
+        ```py
+        >>> from diffusers import StableDiffusionPipeline, StableDiffusionSAGPipeline
+
+        >>> pipe = StableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5")
+        >>> new_pipe = StableDiffusionSAGPipeline.from_pipe(pipe)
+        ```
+        """
+
+        pipeline_is_offloaded = True if hasattr(pipeline, "_all_hooks") and len(pipeline._all_hooks) > 0 else False
+        if pipeline_is_offloaded:
+            # `enable_model_cpu_offload` has be called on the pipeline, offload model and remove hook from model
+            for hook in pipeline._all_hooks:
+                # offload model and remove hook from model
+                hook.offload()
+                hook.remove()
+
+        original_config = dict(pipeline.config)
+        torch_dtype = kwargs.pop("torch_dtype", None)
+
+        # derive the pipeline class to instantiate
+        custom_pipeline = kwargs.pop("custom_pipeline", None)
+        custom_revision = kwargs.pop("custom_revision", None)
+
+        if custom_pipeline is not None:
+            pipeline_class = _get_custom_pipeline_class(custom_pipeline, revision=custom_revision)
+        else:
+            pipeline_class = cls
+
+        expected_modules, optional_kwargs = cls._get_signature_keys(pipeline_class)
+        # true_optional_modules are optional components with default value in signature so it is ok not to pass them to `__init__`
+        # e.g. `image_encoder` for StableDiffusionPipeline
+        true_optional_modules = set(
+            {
+                k
+                for k, v in inspect.signature(cls.__init__).parameters.items()
+                if v.default != inspect._empty and k in expected_modules
+            }
+        )
+
+        signature_types = pipeline_class._get_signature_types()
+
+        pretrained_model_name_or_path = original_config.pop("_name_or_path", None)
+        # allow users pass modules in `kwargs` to override the original pipeline's components
+        passed_class_obj = {k: kwargs.pop(k) for k in expected_modules if k in kwargs}
+
+        original_class_obj = {}
+        for k, v in pipeline.components.items():
+            if k in expected_modules and k not in passed_class_obj:
+                if (
+                    not isinstance(v, torch.nn.Module)
+                    or type(v) in signature_types[k]
+                    or (v is None and k in cls._optional_components)
+                ):
+                    original_class_obj[k] = v
+                else:
+                    logger.warn(
+                        f"component {k} is not switched over to new pipeline because type does not match the expected."
+                        f" {k} is {type(v)} while the new pipeline expect {signature_types[k]}."
+                        f" please pass the correct type of component to the new pipeline. `from_pipe(..., {k}={k})`"
```

Comment:
```suggestion
                        f"component {name} is not switched over to new pipeline because type does not match the expected."
                        f" {name} is {type(component)} while the new pipeline expect {signature_types[name]}."
                        f" please pass the correct type of component to the new pipeline. `from_pipe(..., {name}={name})`"
```